pkg/instrumentation/middleware.go

```diff
@@ -0,0 +1,60 @@
+package instrumentation
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/labstack/echo/v4"
+	echo_middleware "github.com/labstack/echo/v4/middleware"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type MetricsConfig struct {
+	Skipper echo_middleware.Skipper
+	Metrics *Metrics
+}
+
+var defaultConfig MetricsConfig = MetricsConfig{
+	Skipper: echo_middleware.DefaultSkipper,
+	Metrics: NewMetrics(prometheus.NewRegistry()),
+}
+
+// https://github.com/labstack/echo/pull/1502/files
+// This method exist for v5 echo framework
+func matchedRoute(ctx echo.Context) string {
+	pathx := ctx.Path()
+	for _, r := range ctx.Echo().Routes() {
+		if pathx == r.Path {
+			return r.Path
+		}
+	}
+	return ""
+}
+
+func MetricsMiddlewareWithConfig(config *MetricsConfig) echo.MiddlewareFunc {
+	if config == nil {
+		config = &defaultConfig
+	}
+	if config.Skipper == nil {
+		config.Skipper = echo_middleware.DefaultSkipper
+	}
+	if config.Metrics == nil {
+		panic("config.Metrics can not be nil")
+	}
+	return func(next echo.HandlerFunc) echo.HandlerFunc {
+		return func(ctx echo.Context) error {
+			start := time.Now()
+			method := ctx.Request().Method
+			path := matchedRoute(ctx)
+
+			config.Metrics.HttpTotalRequests.Inc()
+			err := next(ctx)
+			if ctx.Response().Status >= 500 {
+				config.Metrics.HttpTotalFailedRequests.Inc()
+			}
+			status := fmt.Sprintf("%d", ctx.Response().Status)
+			config.Metrics.HttpRequestLatency.WithLabelValues(status, method, path).Observe(float64(time.Since(start).Nanoseconds()) / 1000000)
```

Comment:
Check my math, but I think this is milliseconds (1 nanosecond = 1e-6 ms). But the metric is seconds. [ObserveDuration](https://pkg.go.dev/github.com/prometheus/client_golang/prometheus#Timer.ObserveDuration) would return seconds.

Or there is a Milliseconds [function](https://pkg.go.dev/time#Duration.Milliseconds) and we could use milliseconds for the metric.

Reply:
:+1: Nice catch! I have to divide by 1000000000.

I have just realized that the `.Seconds()` method is implemented like the below:

```go
func (d Duration) Seconds() float64 {
	sec := d / Second
	nsec := d % Second
	return float64(sec) + float64(nsec)/1e9
}
```

So I could simplify the expression with just:

```go
config.Metrics.
    HttpRequestLatency.
    WithLabelValues(status, method, path).
    Observe(time.Since(start).Seconds())
```